It should be possible to select a team for a new user during their creation

adishy opened the pull request:
### Status
**READY**

### Description
* When creating a new user as an admin, it should be possible to add the new user to a team if the course has teams.
* Now a dropdown has been added to the /users/new page under the course section that lists all the teams in the course if the course has teams. If the new user is a student, they are added to the list of the team's students. If the user is a GSI or professor they are added to the list of the team's leaders.
* Also, now deleting a staff member will also remove them from their team

### Migrations
NO

### Steps to Test or Reproduce
1. As an admin, create a new user by visiting /users/new in a course with teams
2. Select a team for the user by selecting a team from the dropdown. When the user is created they are a member of the team. If no options from the dropdown are selected, the new user is not added to any team.

### Impacted Areas in Application
* Adding a new user (/users/new)
* controllers/users_controller.rb
* The course view when creating a new user (/views/users/_courses.html.haml)
* The model course_membership.rb to add an attr_accessor called team_in_course to store the new user's team in the course
* courses_helper.rb now includes a helper function to list all the names of the teams

======================
Closes #4308, #4317

maskib commented:
This works well! I would make two changes:
-  The header above the dropdown menu used to select the section to add users to says "select team" but it should say "select ("term for "section")"
- This dropdown menu also now exists on the "edit user" pages. I think that's fine, but if you try to change which section users are in from that page and click "update user", it doesn't actually change the section they're in. The "New Staff Member (x) was successfully updated!" still shows up after that too so people will think that they have in fact changed that user's section.

I'm also wondering if it's worth changing the term "ta" on the "sections" page to something more like "section leader" to be more general? Not sure if that matters, that's just a thought. Thank you!!

maskib commented:
@adishy I'll also just add that I recognize that in many places the terms "sections" and "teams" are used interchangeably, and that we might want to standardize this across the board in the future. What users typically see is "section" or whatever term they're using for "section", so I think we should stick with that for users. Thanks again!

adishy commented:
@maskib Great catch on both the term_for point as well as about the edit user page! I've updated the title to be the singular version of whatever the term for teams are in the course and changed the page so that the "Select Team" dropdown only appears while creating a user (i.e. /users/new) and not when editing an existing user.

adishy commented:
@maksib The requested changes are now on staging!

maskib commented:
@adishy @ebarroso7 Great, thanks! I want to ask: what happens if somebody adds a new staff member to the wrong section in the course or if the staff member needs to change sections midway through the course? Would we have to remove them from the course and re-add them? Possibly not an  issue, I just want to confirm whether this is the case, and if we're okay with that.

adishy commented:
@maskib Currently the interaction for changing the section of an existing staff member is as follows:

1. Visit the /teams page as admin and choose the team the staff member is currently in and edit the team. Uncheck the staff member's name from the "Leaders" section.
2. Go back to the /teams page and choose the new team for the staff member and edit the team. Check the staff member's name in the "Leaders" section for the new team. Note that it is possible add one staff member as the leader for multiple teams.

I do agree that this interaction is a bit tedious :/

maskib commented:
@adishy Got it, thanks! It's a little tedious, but I can't imagine it would have to happen often. Thanks!

ebarroso7 commented on app/controllers/course_memberships_controller.rb:
```diff
@@ -56,10 +56,31 @@ def reactivate
     course_membership = current_course.course_memberships.find(params[:id])
-    Services::CancelsCourseMembership.call course_membership
+    
+    remove_leaders_from_teams(course_membership)
```
I think it would make the logic a little easier to read in this area if you pulled out the conditional check that the user is a staff member from within the `remove_leaders_from_team` to within the destroy function.

so within the `destroy` method it would be: 

if course_membership.user.is_staff?(course_membership.course)
      remove_leaders_from_teams(course_membership)
end

adishy replied:
I believe that particular change was made to keep codeclimate happy, I do agree it was a bit unreadable so it has been updated.

ebarroso7 commented on app/controllers/course_memberships_controller.rb:
```diff
@@ -56,10 +56,31 @@ def reactivate
+      team.save
+    end
+
+    return true
```
Not sure if you need this `return true` here ?

adishy replied:
Updated the function to not include this.

ebarroso7 commented on app/helpers/courses_helper.rb:
```diff
@@ -18,6 +18,11 @@ def available_roles(course)
+  def available_teams(course)
+    teams = course.teams.pluck("name")
+    return teams
+  end
```
with ruby's implicit returns, I think you can shorten this to just `course.teams.pluck("name")`  
not setting or returning `teams`
